sqlite3.go

```diff
@@ -1901,16 +1895,15 @@ func (s *SQLiteStmt) Close() error {
 		return nil
 	}
 	s.closed = true
-	if !s.c.dbConnOpen() {
-		return errors.New("sqlite statement with already closed database connection")
-	}
-	rv := C.sqlite3_finalize(s.s)
+	conn := s.c
+	stmt := s.s
+	s.c = nil
 	s.s = nil
+	runtime.SetFinalizer(s, nil)
+	rv := C.sqlite3_finalize(stmt)
 	if rv != C.SQLITE_OK {
-		return s.c.lastError()
+		return conn.lastError()
```

Comment:
Here too, this may not be generally safe. If the conn was already closed, then `lastError()` is ill-defined.

Personally, I think this is indicative of a deeper design flaw around this library's error handling. It's kind of silly that we get an error code, then drop it on the floor so we can call `sqlite3_errcode` instead. Even the extended error code would be simpler if we just configured the conn to give it back in the first place.

I'm going to file another issue to redo the error handling to avoid these issues. But for now, I think you'll have to check if the conn was already closed in this case, and if so directly construct the `Error` instead of calling `lastError`.

Reply:
> Here too, this may not be generally safe. If the conn was already closed, then lastError() is ill-defined.

It's fine to call `lastError()` here since the db connection may still be open and the error could be due to something else (which we'd want to capture). Additionally, we know that the statement has not been closed before and we [have a test](https://github.com/charlievieth/go-sqlite3/blob/69c42ee761eebcbfe308ce6b13360e4ad4e64512/sqlite3_test.go#L341-L348) that covers the case of this being called with a closed database.

> Personally, I think this is indicative of a deeper design flaw around this library's error handling. It's kind of silly that we get an error code, then drop it on the floor so we can call sqlite3_errcode instead. Even the extended error code would be simpler if we just configured the conn to give it back in the first place.

Agreed, but this would be a breaking change since users might be asserting on `sqlite3.Error.Code`, which we wouldn't have if we enabled extended error code with: `sqlite3_extended_result_codes`.

Reply:
> the db connection may still be open

In `Close`, we are setting `c.db = nil`. That means that `c.lastError()` will no longer work as expected, because it no longer has a reference to the database handle.

> Agreed, but this would be a breaking change since users might be asserting on sqlite3.Error.Code

The extended error code is the same as the regular error code, just with more bits. So this is addressed with a bitmask. See https://www.sqlite.org/rescode.html#primary_result_codes_versus_extended_result_codes

Reply:
Actually, there is something here, but it's a bit different. If the conn is already closed then `SQLiteConn.lastError()` will return nil since we set the conn to nil in `SQLiteConn.Close` so this is safe, but it means that we drop the error from `sqlite3_finalize` when the db conn is closed. We previously returned the error `"sqlite statement with already closed database connection"` if db conn was closed, but removed that check since it was preventing us from calling `sqlite3_finalize` on the statement.

The fix here is to either preserve the error returned by `sqlite3_finalize` if the database is already closed or if `sqlite3_finalize` does return an error we should return the current `"sqlite statement with already closed database connection"` error. Assuming no one is asserting on that error message the later option might be better since it comes directly from sqlite3.

Reply:
Bit of a race condition on us both commenting, but I think my last comment addresses your first point about Close. As for the extended error code - thank you - I forgot that you can use a bitmask for them.

Reply:
@rittneje After some thought I think you're right and that for this change to work the overall error handling needs to be improved. In commit https://github.com/mattn/go-sqlite3/pull/1303/commits/d3deb2d082a5c1adf8283dc744a5a140ef690de7 I updated the error handling to: always use the extended result code; respect the original/offending result code; and safely handle the underlying db connection being closed in `lastError`. It also, changes the logic so that when `sqlite3_close_v2` fails we don't rely on the db (which might be in a weird state) to generate the `sqlite3.Error`.

Thank you again for your review.